pimcore/models/Tool/CustomReport/Config.php

```diff
@@ -150,7 +150,23 @@ public static function getReportsList()
     public static function getAdapter($configuration, $fullConfig = null)
     {
         $type = $configuration->type ? ucfirst($configuration->type) : 'Sql';
-        $adapter = "\\Pimcore\\Model\\Tool\\CustomReport\\Adapter\\{$type}";
+        if ($type === 'Custom') {
+            if (@\class_exists($configuration->className, true)) { // can need autoloader, can send warning and break json response so we use @
+                $adapter = $configuration->className;
+                $adpaterInstance = new $adapter($configuration, $fullConfig);
+                // we make some check for security reason (because className can be choosen at the GUI level)
+                if (get_parent_class($adpaterInstance) !== 'Pimcore\\Model\\Tool\\CustomReport\\Adapter\\AbstractAdapter') {
+                    throw new \Exception("Class" . $configuration->className ." doesn't extend AbstractAdapter");
+                }
+                if ((!method_exists($adpaterInstance, 'getData')) || (!method_exists($adpaterInstance, 'getColumns')) || (!method_exists($adpaterInstance, 'getAvailableOptions'))) {
```

Comment:
As my previous line comment, it was for security purpose too, but my code is not good, as it will alwyas send true (if class extend AbstractAdapter). In ideal, the custom reports should maybe "implement" an interface than extend the abstract class, but it will need more change in the core code.
So you are right you can remove this check too.

Or if you prefer to keep it, it need to check it exist for the class itself (not including the parent), using reflection could maybe do the trick:
$reflectionMethod = new \ReflectionMethod($adapterInstance, 'getData');
$isNotDeclared = ($reflectionMethod->getDeclaringClass()->getName() !== get_class($adpaterInstance));
This way we are sure methods are implemented. We could have use instead isAbstract() method of ReflectionMethod to ensure it has been implemented too.